File: frontend/packages/contracts/src/strategy.ts
import { z } from "zod";
import {
  ExchangeSchema,
  IdSchema,
  SegmentSchema,
  TimeframeSchema,
  UtcDateTimeSchema,
} from "./common";
import { IndicatorNameSchema, checkIndicatorParams } from "./indicators";

export const PriceFieldSchema = z.enum(["open", "high", "low", "close", "volume"]);
export type PriceField = z.infer<typeof PriceFieldSchema>;

export const OffsetSchema = z.number().int().min(0).max(500);

export const OperandPriceSchema = z.strictObject({
  kind: z.literal("price"),
  field: PriceFieldSchema,
  /** Bars ago (D51); absent = 0. */
  offset: OffsetSchema.optional(),
});
export type OperandPrice = z.infer<typeof OperandPriceSchema>;

export const OperandIndicatorSchema = z.strictObject({
  kind: z.literal("indicator"),
  name: IndicatorNameSchema,
  params: z.record(z.string(), z.number()),
  /** Bars ago (D51); absent = 0. */
  offset: OffsetSchema.optional(),
});
export type OperandIndicator = z.infer<typeof OperandIndicatorSchema>;

export const OperandNumberSchema = z.strictObject({
  kind: z.literal("number"),
  value: z.number(),
});
export type OperandNumber = z.infer<typeof OperandNumberSchema>;

export const OperandSchema = z.discriminatedUnion("kind", [
  OperandPriceSchema,
  OperandIndicatorSchema,
  OperandNumberSchema,
]);
export type Operand = z.infer<typeof OperandSchema>;

export const ConditionOpSchema = z.enum([
  "crosses_above",
  "crosses_below",
  "gt",
  "gte",
  "lt",
  "lte",
  "eq",
]);
export type ConditionOp = z.infer<typeof ConditionOpSchema>;

export const ConditionSchema = z.strictObject({
  left: OperandSchema,
  op: ConditionOpSchema,
  right: OperandSchema,
});
export type Condition = z.infer<typeof ConditionSchema>;

export const RuleGroupCombinatorSchema = z.enum(["all", "any"]);
export type RuleGroupCombinator = z.infer<typeof RuleGroupCombinatorSchema>;

export const RuleGroupSchema = z.strictObject({
  combinator: RuleGroupCombinatorSchema,
  conditions: z.array(ConditionSchema).min(1),
});
export type RuleGroup = z.infer<typeof RuleGroupSchema>;

export const IndexNameSchema = z.enum(["NIFTY 50", "NIFTY BANK", "NIFTY NEXT 50"]);
export type IndexName = z.infer<typeof IndexNameSchema>;

export const UniverseSymbolsSchema = z.strictObject({
  type: z.literal("symbols"),
  symbols: z.array(z.string().min(1)).min(1),
});
export type UniverseSymbols = z.infer<typeof UniverseSymbolsSchema>;

export const UniverseIndexSchema = z.strictObject({
  type: z.literal("index"),
  index: IndexNameSchema,
});
export type UniverseIndex = z.infer<typeof UniverseIndexSchema>;

export const UniverseSchema = z.discriminatedUnion("type", [
  UniverseSymbolsSchema,
  UniverseIndexSchema,
]);
export type Universe = z.infer<typeof UniverseSchema>;

export const SizingFixedQtySchema = z.strictObject({
  type: z.literal("fixed_qty"),
  qty: z.number().int().positive(),
});
export type SizingFixedQty = z.infer<typeof SizingFixedQtySchema>;

export const SizingFixedAmountSchema = z.strictObject({
  type: z.literal("fixed_amount"),
  amountPaise: z.number().int().positive(),
});
export type SizingFixedAmount = z.infer<typeof SizingFixedAmountSchema>;

export const SizingPercentEquitySchema = z.strictObject({
  type: z.literal("percent_equity"),
  percent: z.number().gt(0).lte(100),
});
export type SizingPercentEquity = z.infer<typeof SizingPercentEquitySchema>;

export const SizingSchema = z.discriminatedUnion("type", [
  SizingFixedQtySchema,
  SizingFixedAmountSchema,
  SizingPercentEquitySchema,
]);
export type Sizing = z.infer<typeof SizingSchema>;

export const RiskSchema = z.strictObject({
  stopLossPercent: z.number().positive().nullable(),
  targetPercent: z.number().positive().nullable(),
});
export type Risk = z.infer<typeof RiskSchema>;

/** Cost averaging (D53): buy again each `dropPercent` fall below the last buy, up to `maxAdds` times. */
export const AveragingSchema = z.strictObject({
  dropPercent: z.number().gt(0).lte(50),
  maxAdds: z.number().int().min(1).max(10),
});
export type Averaging = z.infer<typeof AveragingSchema>;

/** A strategy is rules only (D25); symbols are chosen per backtest run (`BacktestRun.universe`). */
const baseSpecFields = {
  segment: SegmentSchema,
  exchange: ExchangeSchema,
  timeframe: TimeframeSchema,
  sizing: SizingSchema,
  risk: RiskSchema,
  /** Absent = off (D53). */
  averaging: AveragingSchema.optional(),
};

export const StrategySpecVisualSchema = z.strictObject({
  mode: z.literal("visual"),
  ...baseSpecFields,
  entry: RuleGroupSchema,
  exit: RuleGroupSchema,
});
export type StrategySpecVisual = z.infer<typeof StrategySpecVisualSchema>;

export const StrategySpecPythonSchema = z.strictObject({
  mode: z.literal("python"),
  ...baseSpecFields,
  code: z.string().min(1),
});
export type StrategySpecPython = z.infer<typeof StrategySpecPythonSchema>;

export const StrategySpecSchema = z.discriminatedUnion("mode", [
  StrategySpecVisualSchema,
  StrategySpecPythonSchema,
]);
export type StrategySpec = z.infer<typeof StrategySpecSchema>;

export const StrategyVersionSchema = z.strictObject({
  version: z.number().int().min(1),
  createdAt: UtcDateTimeSchema,
  note: z.string(),
  spec: StrategySpecSchema,
});
export type StrategyVersion = z.infer<typeof StrategyVersionSchema>;

export const StrategyStatusSchema = z.enum(["draft", "active", "archived"]);
export type StrategyStatus = z.infer<typeof StrategyStatusSchema>;

export const StrategySchema = z
  .strictObject({
    id: IdSchema,
    name: z.string().min(1),
    description: z.string(),
    status: StrategyStatusSchema,
    latestVersion: z.number().int().min(1),
    versions: z.array(StrategyVersionSchema).min(1),
    createdAt: UtcDateTimeSchema,
    updatedAt: UtcDateTimeSchema,
  })
  .refine(
    (data) => {
      const maxVersion = Math.max(...data.versions.map((v) => v.version));
      return data.latestVersion === maxVersion;
    },
    {
      message: "latestVersion must equal the maximum version in versions array",
      path: ["latestVersion"],
    },
  );
export type Strategy = z.infer<typeof StrategySchema>;

/** Indicator settings problems in a visual spec (D51); write bodies refuse them, reads stay tolerant. */
export function specParamProblems(spec: StrategySpec): string[] {
  if (spec.mode !== "visual") return [];
  const operands = [spec.entry, spec.exit].flatMap((g) =>
    g.conditions.flatMap((c) => [c.left, c.right]),
  );
  return operands.flatMap((o) =>
    o.kind === "indicator" ? checkIndicatorParams(o.name, o.params) : [],
  );
}

const checkedSpec = (body: { spec: StrategySpec }, ctx: z.RefinementCtx) => {
  for (const message of specParamProblems(body.spec)) {
    ctx.addIssue({ code: "custom", message, path: ["spec"] });
  }
};

/** Body of `POST /strategies` (D43): creates version 1 of a `draft` strategy. */
export const StrategyCreateSchema = z
  .strictObject({
    name: z.string().min(1),
    description: z.string(),
    spec: StrategySpecSchema,
  })
  .superRefine(checkedSpec);
export type StrategyCreate = z.infer<typeof StrategyCreateSchema>;

/** Body of `POST /strategies/{id}/versions` (D43): versions are immutable; this adds latest + 1. */
export const StrategyVersionCreateSchema = z
  .strictObject({
    note: z.string(),
    spec: StrategySpecSchema,
  })
  .superRefine(checkedSpec);
export type StrategyVersionCreate = z.infer<typeof StrategyVersionCreateSchema>;

/** Body of `PATCH /strategies/{id}` (D43): at least one field. */
export const StrategyUpdateSchema = z
  .strictObject({
    name: z.string().min(1).optional(),
    description: z.string().optional(),
    status: StrategyStatusSchema.optional(),
  })
  .refine((u) => Object.keys(u).length > 0, { message: "Change at least one field" });
export type StrategyUpdate = z.infer<typeof StrategyUpdateSchema>;

File: backend/libs/nova_contracts/src/nova_contracts/strategy.py
"""Strategy contracts: mirror `frontend/packages/contracts/src/strategy.ts` (D9, D25, D43)."""

from typing import Annotated, Literal, Self

from pydantic import Field, model_validator

from nova_contracts.common import Contract, Exchange, Id, Segment, Timeframe, UtcDateTime
from nova_contracts.indicators import IndicatorName, param_problems

PriceField = Literal["open", "high", "low", "close", "volume"]
ConditionOp = Literal["crosses_above", "crosses_below", "gt", "gte", "lt", "lte", "eq"]
StrategyStatus = Literal["draft", "active", "archived"]
NonEmpty = Annotated[str, Field(min_length=1)]
# Bars ago (D51): absent on the wire = 0, and 0 is never written back (old specs stay identical).
Offset = Annotated[int, Field(ge=0, le=500, exclude_if=lambda v: v == 0)]


class OperandPrice(Contract):
    kind: Literal["price"]
    field: PriceField
    offset: Offset = 0


class OperandIndicator(Contract):
    kind: Literal["indicator"]
    name: IndicatorName
    params: dict[str, float]
    offset: Offset = 0


class OperandNumber(Contract):
    kind: Literal["number"]
    value: float


Operand = Annotated[OperandPrice | OperandIndicator | OperandNumber, Field(discriminator="kind")]


class Condition(Contract):
    left: Operand
    op: ConditionOp
    right: Operand


class RuleGroup(Contract):
    combinator: Literal["all", "any"]
    conditions: Annotated[list[Condition], Field(min_length=1)]


class SizingFixedQty(Contract):
    type: Literal["fixed_qty"]
    qty: Annotated[int, Field(gt=0)]


class SizingFixedAmount(Contract):
    type: Literal["fixed_amount"]
    amount_paise: Annotated[int, Field(gt=0)]


class SizingPercentEquity(Contract):
    type: Literal["percent_equity"]
    percent: Annotated[float, Field(gt=0, le=100)]


Sizing = Annotated[
    SizingFixedQty | SizingFixedAmount | SizingPercentEquity, Field(discriminator="type")
]


class Risk(Contract):
    stop_loss_percent: Annotated[float, Field(gt=0)] | None
    target_percent: Annotated[float, Field(gt=0)] | None


class Averaging(Contract):
    """Cost averaging (D53): buy again each `drop_percent` fall below the last buy."""

    drop_percent: Annotated[float, Field(gt=0, le=50)]
    max_adds: Annotated[int, Field(ge=1, le=10)]


class _SpecBase(Contract):
    segment: Segment
    exchange: Exchange
    timeframe: Timeframe
    sizing: Sizing
    risk: Risk
    # Absent = off, and never written when off, so older specs stay identical (D53).
    averaging: Annotated[Averaging | None, Field(exclude_if=lambda v: v is None)] = None


class StrategySpecVisual(_SpecBase):
    mode: Literal["visual"]
    entry: RuleGroup
    exit: RuleGroup


class StrategySpecPython(_SpecBase):
    mode: Literal["python"]
    code: NonEmpty


StrategySpec = Annotated[StrategySpecVisual | StrategySpecPython, Field(discriminator="mode")]


class StrategyVersion(Contract):
    version: Annotated[int, Field(ge=1)]
    created_at: UtcDateTime
    note: str
    spec: StrategySpec


class Strategy(Contract):
    id: Id
    name: NonEmpty
    description: str
    status: StrategyStatus
    latest_version: Annotated[int, Field(ge=1)]
    versions: Annotated[list[StrategyVersion], Field(min_length=1)]
    created_at: UtcDateTime
    updated_at: UtcDateTime

    @model_validator(mode="after")
    def _latest(self) -> Self:
        if self.latest_version != max(v.version for v in self.versions):
            raise ValueError("latestVersion must equal the maximum version in versions")
        return self


def spec_param_problems(spec: StrategySpecVisual | StrategySpecPython) -> list[str]:
    """Indicator settings problems of a visual spec (D51): writes refuse them, reads do not."""
    if not isinstance(spec, StrategySpecVisual):
        return []
    operands = [o for g in (spec.entry, spec.exit) for c in g.conditions for o in (c.left, c.right)]
    return [
        problem
        for o in operands
        if isinstance(o, OperandIndicator)
        for problem in param_problems(o.name, o.params)
    ]


class _CheckedSpec(Contract):
    spec: StrategySpec

    @model_validator(mode="after")
    def _params(self) -> Self:
        problems = spec_param_problems(self.spec)
        if problems:
            raise ValueError("; ".join(problems))
        return self


class StrategyCreate(_CheckedSpec):
    name: NonEmpty
    description: str


class StrategyVersionCreate(_CheckedSpec):
    note: str


class StrategyUpdate(Contract):
    name: NonEmpty | None = None
    description: str | None = None
    status: StrategyStatus | None = None

    @model_validator(mode="after")
    def _something(self) -> Self:
        if not self.model_fields_set:
            raise ValueError("Change at least one field")
        return self
